ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/TransactionSimulator.java

```diff
@@ -89,40 +90,43 @@ public Optional<TransactionSimulatorResult> process(
       final OperationTracer operationTracer,
       final long blockNumber) {
     final BlockHeader header = blockchain.getBlockHeader(blockNumber).orElse(null);
-    return process(callParams, transactionValidationParams, operationTracer, header);
-  }
-
-  public Optional<TransactionSimulatorResult> process(
-      final CallParameter callParams, final Hash blockHeaderHash) {
-    final BlockHeader header = blockchain.getBlockHeader(blockHeaderHash).orElse(null);
     return process(
         callParams,
-        TransactionValidationParams.transactionSimulator(),
-        OperationTracer.NO_TRACING,
+        transactionValidationParams,
+        operationTracer,
+        (mutableWorldState, transactionSimulatorResult) -> transactionSimulatorResult,
         header);
   }
 
-  public Optional<TransactionSimulatorResult> process(
-      final CallParameter callParams, final long blockNumber) {
+  public <U> Optional<TransactionSimulatorResult> processAtHead(final CallParameter callParams) {
     return process(
         callParams,
         TransactionValidationParams.transactionSimulator(),
         OperationTracer.NO_TRACING,
-        blockNumber);
+        (mutableWorldState, transactionSimulatorResult) -> transactionSimulatorResult,
+        blockchain.getChainHeadHeader());
   }
 
-  public Optional<TransactionSimulatorResult> processAtHead(final CallParameter callParams) {
+  public <U> Optional<U> processAtHead(
+      final CallParameter callParams,
+      final BiFunction<
+              MutableWorldState, Optional<TransactionSimulatorResult>, ? extends Optional<U>>
+          mapper) {
     return process(
         callParams,
         TransactionValidationParams.transactionSimulator(),
         OperationTracer.NO_TRACING,
+        mapper,
         blockchain.getChainHeadHeader());
   }
 
-  public Optional<TransactionSimulatorResult> process(
+  public <U> Optional<U> process(
       final CallParameter callParams,
       final TransactionValidationParams transactionValidationParams,
       final OperationTracer operationTracer,
+      final BiFunction<
+              MutableWorldState, Optional<TransactionSimulatorResult>, ? extends Optional<U>>
+          mapper,
```

Comment:
This is a pretty gnarly function type - can this be any clearer, maybe extracting into FunctionalInterface with a good name? A comment at least? 

I currently have no idea what different classes it can return and the variable name `mapper` provides no clues.

Reply:
> > Additionally, the PR addresses a side effect associated with this fix that impacts eth_call.
> 
> What was the side-effect?

it was not ethcall it was traceCall

In this part we are trying to get the account from the DB [https://github.com/hyperledger/besu/blob/main/ethereum/api/src/main/java/org/hyper[…]i/jsonrpc/internal/results/tracing/diff/StateDiffGenerator.java](https://github.com/hyperledger/besu/blob/main/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/diff/StateDiffGenerator.java#L48) before this PR it was from the cache. The problem in this case the worldstate is already close at this moment [https://github.com/hyperledger/besu/blob/main/ethereum/api/src/main/java/org/hyper[…]edger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java](https://github.com/hyperledger/besu/blob/main/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java#L105)

Reply:
Done hope it will b clear like that

Reply:
https://github.com/hyperledger/besu/pull/5330/files#diff-5bf48e64343e7cc59a7ded7bbaaa71872d71b012172d763d5ee5baa1070fa37c

Reply:
so IIUC, it was working before because data survives in the cache even after the worldstate is closed?

Reply:
yes you got it